### DISCUSSION TITLE
does disabling Unicode mean I can't search for \u2603?

### DISCUSSION THREAD

**Discussion by d3lm** [rust-lang/regex#813] [Q&A] [answered]
Hey there 👋 

Is it correct that if I disable all `unicode` features that a regex won't match any unicode at all? I am a bit confused when I read the documentation.

In my use case, I want to apply a regex on a byte slice that represents some file content and that slice could potentially also contain unicode characters. If I were to disable unicode and search for say `\u2603` in a file that contains those bytes (☃), does that mean it wouldn't find it?

Basically I am always operating on `&[u8]` so I guess I can use the `bytes` submodule. But I am wondering if I can disable unicode because binary size is very important.

**Comment by BurntSushi** [comment, marked as answer]
> I am a bit confused when I read the documentation.

To be clear, and to make sure we're talking about the same thing, these are the docs that I think should answer this question: https://docs.rs/regex/1.5.4/regex/#crate-features

In particular, the last couple sentences of the second paragraph I believe are crucial here. Do they answer your question? If not, there might be some other misunderstanding (perhaps on my part) going on here:

> Stated differently, enabling or disabling any of the features below can only add or subtract from the total set of valid regular expressions. Enabling or disabling a feature will never modify the match semantics of a regular expression.

So in other words, if compiling a regex like `\u2603` works with all of the Unicode features disabled, then it is guaranteed to have the same semantics as it would if any of those features were enabled.

Now maybe you're talking about the `u` regex flag, in which case, if you try matching `\u2603` with `u` disabled, then you should get a regex compilation error. But the `u` regex flag is unrelated to binary size since it's a runtime option, so I'm guessing this is not what you meant.

Also,

>  so I guess I can use the bytes submodule

When you say, "so I guess I can use...," what do you mean here? If you're matching on `&[u8]`, then the `bytes` sub-module API is exactly the right thing you want to be using, and this is (or should be) orthogonal from the rest of your question. The `bytes` API has all of the Unicode features of the main top-level module API.

Also, you speak of "disabling Unicode," but I think what you mean is, "disable all crate features starting with `unicode`." If you take a look at those features, none of them correspond to whether you can search for a Unicode codepoint. And in particular, searching for a Unicode codepoint doesn't require any Unicode tables, and thus, there is no binary size impact nor is there any reason to disable such functionality at compile time (AFAIK).

> binary size is very important

You'll likely also want to disable the `perf` features.

**Reply by d3lm** [reply]
Thanks so much for your reply! This was very helpful. I think there was some misunderstanding on my end because I actually thought that searching for unicode codepoints would require those unicode tables.

And yes, by "disabling unicode" I actually meant disabling all the `unicode*` features.

I think you have answered my question pretty much. To summarize, making sure I understood correctly, searching for unicode codepoints does not require the unicode tables and I can only compile a regex which contains unicode if I also enable the `u` flag. But all of this works without enabling the `unicode*` features. Is that right?

> You'll likely also want to disable the perf features.

For some reason, disabling that feature has no effect on the binary size.

**Reply by BurntSushi** [reply]
> because I actually thought that searching for unicode codepoints would require those unicode tables

Aye. The only thing required is to map a codepoint to its UTF-8 encoding. That UTF-8 encoding is then built into the finite automata of the regex engine. So in the case of `\u2603`, its UTF-8 encoding is `\xE2\x98\x83`. Its corresponding (anchored) DFA is thus:

```
$ regex-cli debug dfa dense '\u2603' -aC
            parse time:  10.181µs
        translate time:  1.643µs
      compile nfa time:  173.901µs
            nfa memory:  444
compile dense dfa time:  89.611µs
      dense dfa memory:  14368
 dense alphabet length:  257
          dense stride:  512

dense::DFA(
D 000000:
Q 000001:
 *000002:
 >000003: \xE2 => 6
  000004: \x83 => 5
  000005: \x00-\xFF => 2, EOI => 2
  000006: \x98 => 4

START-GROUP(ALL)
  NonWordByte => 000003
  WordByte => 000003
  Text => 000003
  Line => 000003
state count: 7
pattern count: 1
)
```

And that's pretty much it. No tables necessary. :-)

> To summarize, making sure I understood correctly, searching for unicode codepoints does not require the unicode tables and I can only compile a regex which contains unicode if I also enable the `u` flag. But all of this works without enabling the `unicode*` features. Is that right?

Correct.

> For some reason, disabling that feature has no effect on the binary size.

It removes _a lot_ of code. So I suspect there is something amiss with your measurement procedure. Here's the analysis I did (including documenting my measurement procedure) when I added those features and which shows that removing `perf` features decreases binary size: https://github.com/rust-lang/regex/pull/613#issue-488338216

**Reply by d3lm** [reply]
Thanks so much for your in-depth response! That is very helpful 🙏 

I ll have another look at the `perf` feature. I am actually using your `globset` which includes `regex` and what I did is I added `regex` as a dependency to my own crate so I can control the features.

Basically what I do is:
```
globset = "0.4.8"
regex = { version = "1.1.5", features = ["std"], default-features = false }
```
So I'd guess that this would only enable `std`. I can see it adds about 300K, but interestingly, if I enable `perf` it doesn't really go up. Maybe I am not using any of it or something?

**Reply by BurntSushi** [reply]
Yeah, it's because [`globset` unconditionally enables the `perf` feature](https://github.com/BurntSushi/ripgrep/blob/ba535fb5a34f6d703fc652e8dfc23c0fed78fc74/crates/globset/Cargo.toml#L27). There is no way to disable it. Since Cargo features can really only be added, if _any_ part of your dependency graph depending on `regex` enables feature `foo`, then it is enabled regardless of anything else you do.

You could file an issue against globset for this, but to be honest:

1. `globset` needs an overhaul.
2. Optimizing for binary size at the level of `globset` is not a goal I'm really interested in spending time on. `globset` has so many dependencies that optimizing for binary size is a fool's errand IMO.

Basically, if binary size is _really_ that important, then you probably shouldn't be using `globset`.

**Reply by d3lm** [reply]
Ok that makes sense. Do you know of any alternative? What I need is to run a set of glob patterns against strings. But most of the binary size from `globset` comes from `regex` and I need `regex` as well, though I could get away with disabling `perf`.

**Reply by d3lm** [reply]
Also, I don't need any iterator over paths, because I use it on an in-memory file system I wrote. So for the globbing, I only want to apply some globs on strings.

**Reply by d3lm** [reply]
I could use https://docs.rs/glob/0.2.11/glob/struct.Pattern.html but then I'd have to execute each glob individually and I think the optimization of globset is pretty cool and if you have a lot of globs, which I would most likely have, then it does make a difference.

**Reply by BurntSushi** [reply]
No, I don't. Especially given that I don't know what problem you're trying to solve or _why_ binary size is "very important." The only place I can conceive of binary size being important is potentially in WASM or in certain embedded contexts. If you're not in either of those, I'd probably urge you to re-evaluate whether binary size is actually important or not.

At that point, you'll need to consider which is more important: the use of globs, performance or binary size. In theory, this is a false choice, but only in theory. In practice, you'll have to choose between those three things (or carving your own path, e.g., by implementing your own globbing facilities in a way that meets your demands).

**Reply by d3lm** [reply]
Yea, binary size is important because I am compiling for the Web therefore WASM. What I am trying to do is to build a simplified version of ripgrep used by VSCode. So I need to be able to glob on file paths (strings) and then search for patterns in the bytes (the file content). It doesn't have to be as optimised as ripgrep, also because it would significantly increase the size of the binary. Also, as mentioned earlier, I am operating on an in memory file system and not a real file system.

**Reply by BurntSushi** [reply]
More practically, I think these are your choices:

1. Recognize that binary size is important, but mark it as a v2 feature. Move forward without caring too much about it.
2. Fork globset and trim it down.
3. Use the `glob` crate and be okay with the glob perf loss. You might be able to get farther here than you think. e.g., I think this is how `ag` implements gitignore logic. `ag` does not have anything like `globset`.
4. Build your own globbing implementation. This is, after all, what I did.

**Reply by d3lm** [reply]
Yea that makes sense. Thanks a lot!
